### PR TITLE
ARIA-9-API-for-operation-context

### PR DESCRIPTION
**Pull request by mxmrlv**

### CODE REVIEW

File: aria/context/__init__.py [hunk 1]
```diff
@@ -18,3 +18,4 @@
 """
 
 from . import workflow, operation
+from .toolbelt import toolbelt
```
Comment (mxmrlv) [inline aria/context/__init__.py]: added

File: aria/context/common.py [hunk 2]
```diff
@@ -130,34 +121,28 @@ def id(self):
...
         return self._id
 
-    def download_blueprint_resource(self, destination, path=None):
+    def download_resource(self, destination, path=None):
         """
         Download a blueprint resource from the resource storage
         """
-        return self.resource.blueprint.download(
+        return_value = self.resource.deployment.download(
             entry_id=self.blueprint.id,
```
Comment (dankilman) [inline aria/context/common.py]: deployment_id

File: aria/context/common.py [hunk 3]
```diff
@@ -130,34 +121,28 @@ def id(self):
...
-
-    def download_deployment_resource(self, destination, path=None):
-        """
-        Download a deployment resource from the resource storage
-        """
-        return self.resource.deployment.download(
-            entry_id=self._deployment_id,
-            destination=destination,
-            path=path)
+        if not path.isdir(destination) or len(os.listdir(destination)) > 0:
```
Comment (dankilman) [inline aria/context/common.py]: storage_error

File: aria/context/toolbelt.py [hunk 4]
```diff
@@ -13,79 +13,67 @@
...
-    """
     @property
     def relationships_to_me(self):
         """
         Any relationship which the current node is the target of
         :return:
         """
         assert isinstance(self._op_context, operation.NodeOperationContext)
-        for node_instance in self._workflow_context.node_instances:
+        node_instances = self._op_context.model.node_instance.iter(
```
Comment (dankilman) [inline aria/context/toolbelt.py]: remove

File: aria/decorators.py [hunk 5]
```diff
@@ -60,10 +60,8 @@ def operation(func=None, toolbelt=False, suffix_template=''):
     @wraps(func)
     def _wrapper(**func_kwargs):
         if toolbelt:
-            with context.toolbelt(func_kwargs.get('ctx')) as operation_toolbelt:
-                func_kwargs.setdefault('toolbelt', operation_toolbelt)
-                validate_function_arguments(func, func_kwargs)
-                return func(**func_kwargs)
+            operation_toolbelt = context.toolbelt(func_kwargs.get('ctx'))
```
Comment (dankilman) [inline aria/decorators.py]: `[]`

File: aria/workflows/api/task.py [hunk 6]
```diff
@@ -86,64 +89,45 @@ def __init__(self,
...
-    """
-    Represents a node based operation
-    """
-    def __init__(self,
-                 actor,
-                 name,
-                 *args,
-                 **kwargs):
+    @classmethod
+    def node_instance(cls, actor, name, inputs=None, *args, **kwargs):
```
Comment (dankilman) [inline aria/workflows/api/task.py]: instance

File: aria/workflows/api/task.py [hunk 7]
```diff
@@ -86,64 +89,45 @@ def __init__(self,
...
+        operation_inputs.update(inputs or {})
+        return cls(name=name,
+                   actor=actor,
+                   operation_mapping=operation_details.get('operation', ''),
+                   inputs=operation_inputs,
+                   *args,
+                   **kwargs)
+
+    @classmethod
+    def relationship_instance(cls, name, actor, operation_end, inputs=None, *args, **kwargs):
```
Comment (dankilman) [inline aria/workflows/api/task.py]: instance

File: aria/workflows/api/task.py [hunk 8]
```diff
@@ -86,64 +89,45 @@ def __init__(self,
...
-        operation_mapping = getattr(actor.relationship, side)[name].get('operation')
-        operation_inputs = getattr(actor.relationship, side)[name].get('inputs', {})
-        operation_inputs.update(kwargs.pop('inputs', {}))
-        super(RelationshipOperationTask, self).__init__(actor=actor,
-                                                        name=name,
-                                                        operation_mapping=operation_mapping,
-                                                        inputs=operation_inputs,
-                                                        *args,
-                                                        **kwargs)
+        operation_details = getattr(actor.relationship, operation_end)[name]
```
Comment (dankilman) [inline aria/workflows/api/task.py]: validate operation_end

File: aria/workflows/api/task.py [hunk 9]
```diff
@@ -86,64 +89,45 @@ def __init__(self,
...
+    @classmethod
+    def relationship_instance(cls, name, actor, operation_end, inputs=None, *args, **kwargs):
         """
         Represents a relationship based operation
 
-        :param actor: the relationship of which this operation belongs to.
+        :param actor: the node of which this operation belongs to.
         :param name: the name of the operation.
-        :param target: whether this is a source or target operation.
+        :param operation_end: source or target end of the relationship
```
Comment (dankilman) [inline aria/workflows/api/task.py]: real names

File: aria/workflows/executor/celery.py [hunk 10]
```diff
@@ -46,7 +46,7 @@ def execute(self, task):
         self._tasks[task.id] = task
         self._results[task.id] = self._app.send_task(
```
Comment (dankilman) [inline aria/workflows/executor/celery.py]: put `ctx` in `inputs`

File: tests/context/test_toolbelt.py [hunk 11]
```diff
@@ -45,13 +44,8 @@ def executor():
         result.close()
 
 
-def test_operation_tool_belt(workflow_context, executor):
-    operation_name = 'aria.interfaces.lifecycle.create'
+def _create_blueprint_in_storate(workflow_context):
```
Comment (dankilman) [inline tests/context/test_toolbelt.py]: fix

File: aria/storage/drivers.py [hunk 12]
```diff
@@ -370,6 +370,8 @@ def data(self, entry_type, entry_id, path=None):
         :rtype: bytes
         """
         resource = os.path.join(self.directory, entry_type, entry_id, path or '')
+        if not os.path.exists(resource):
+            raise StorageError("Resource does not exist")
```
Comment (dankilman) [inline aria/storage/drivers.py]: add resource to message

File: aria/workflows/api/task.py [hunk 13]
```diff
@@ -90,40 +91,45 @@ def __init__(self,
...
-                   actor=actor,
+                   actor=instance,
                    operation_mapping=operation_details.get('operation', ''),
                    inputs=operation_inputs,
                    *args,
                    **kwargs)
 
     @classmethod
-    def relationship_instance(cls, name, actor, operation_end, inputs=None, *args, **kwargs):
+    def relationship_instance(cls, instance, actor, operation_end, inputs=None, *args, **kwargs):
```
Comment (dankilman) [inline aria/workflows/api/task.py]: fix

File: aria/context/common.py [hunk 14]
```diff
@@ -125,24 +128,21 @@ def download_resource(self, destination, path=None):
         """
         Download a blueprint resource from the resource storage
         """
-        return_value = self.resource.deployment.download(
-            entry_id=self.blueprint.id,
-            destination=destination,
-            path=path)
-        if not path.isdir(destination) or len(os.listdir(destination)) > 0:
-            return self.resource.blueprint.download(entry_id=self._deployment_id,
+        try:
```
Comment (dankilman) [inline aria/context/common.py]: add tests

**Pull request merged**